src/gc.c

```diff
@@ -10,79 +10,148 @@
 // will wait for the next GC, to allow the space to be reused more
 // efficiently. default = on.
 #define FREE_PAGES_EAGER
-
 #include <stdlib.h>
 #include <string.h>
+#include <strings.h>
 #include <assert.h>
-#ifdef USE_MMAP
-# include <sys/mman.h>
-# include <malloc.h>
-#endif
 #include "julia.h"
 #include "julia_internal.h"
-
-#ifdef _P64
-# ifdef USE_MMAP
-#  define GC_PAGE_SZ 16384//bytes
-# else
-#  define GC_PAGE_SZ 12288//bytes
-# endif
-#else
-#define GC_PAGE_SZ  8192//bytes
+#ifndef _OS_WINDOWS_
+#include <sys/mman.h>
+#ifdef _OS_DARWIN_
+#define MAP_ANONYMOUS MAP_ANON
+#endif
 #endif
 
 #ifdef __cplusplus
 extern "C" {
 #endif
 
-typedef struct _gcpage_t {
-    char data[GC_PAGE_SZ];
+#pragma pack(push, 1)
+
+typedef struct {
     union {
-        struct _gcpage_t *next;
-        char _pad[8];
+        uintptr_t header;
+        struct {
+            uintptr_t gc_bits:2;
+            uintptr_t pooled:1;
+        };
     };
-} gcpage_t;
+    char data[];
+} buff_t;
 
 typedef struct _gcval_t {
     union {
         struct _gcval_t *next;
         uptrint_t flags;
-        uptrint_t data0;  // overlapped
-        uptrint_t marked:1;
+        uptrint_t gc_bits:2;
     };
 } gcval_t;
 
-typedef struct _pool_t {
-    size_t osize;
-    gcpage_t *pages;
-    gcval_t *freelist;
-} pool_t;
+// layout for small (<2k) objects
+
+#define GC_PAGE_LG2 14 // log2(size of a page)
+#define GC_PAGE_SZ (1 << GC_PAGE_LG2) // 16k
```

Comment:
ah, is this 16k a typo ? because if those really are 64k pages I don't see how the current code can work since it assumes that it can free gc pages individually, which it can't.
The required change should not be too hard, just have to check that the whole 64k chunk is free before madvise.